### PR TITLE
First working implementation of OTR

### PR DESCRIPTION
**Pull request by JocelynDelalande**
You have to start OTR key negociation with a message like `?OTRv24x?`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,6 +33,7 @@
     "lodash": "~2.4.1",
     "mkdirp": "^0.5.0",
     "moment": "~2.7.0",
+    "otr": "^0.2.15",
```
Comment (astorije) [inline package.json:36]: I noticed this package is in [license MPL 2.0](https://www.mozilla.org/en-US/MPL/2.0/). I don't know this one, do you think it could conflict with the rest of Shout?

Comment (JocelynDelalande) [inline package.json:36]: Not an expert, but seems fine to me, what we are doing here is either :
- linking (when distributing shout on source form)
- distribution in larger project (if distributing in a bundled form, but there is no such distribution of shout at the moment as far as I know)

[that wikipedia page](https://en.wikipedia.org/wiki/Comparison_of_free_and_open-source_software_licenses#General_comparison) states that linking to MPL 2.0 code is permissive, all license analysis I read like [this one](http://www.gatsbymedia.net/development-articles/why-we-use-the-mpl) or [this one](https://julien.ponge.org/blog/mozilla-public-license-v2-a-good-middleground/) states that MPL do not contaminates the whole project with MPL copyleft, but only the _files_ of the work covered by MPL (in our case, the OTR lib, and optional change we would operate on it). 

I had to complete my license reading by 3rd-party analysis as raw law-geek-style license text is too vague to me.

What I don't manage to understand is if we should put some legal mention somewhere, but that's not a blocking point. Maybe some license-guru advice would be valuable, but IMHO we already know that we will not have to go backward on this choice for a licensing reason.

**Comment by JocelynDelalande** [comment]
> This file has whitespace issues, see my other comment.

Ok, fixed on both files, well spotted :)

**Comment by JocelynDelalande** [comment]
@astorije cleaned the inheritance thing.

File: package.json (hunk 1 above)
Comment (astorije) [inline package.json:36]: Great, thanks for spending your time on that, sorry about that ^^

:+1:

File: src/client.js [hunk 2]
```diff
@@ -73,8 +78,11 @@ function Client(sockets, name, config) {
 		});
 	}
 }
+util.inherits(Client, EventEmitter);
```
Comment (astorije) [inline src/client.js:82]: [This](https://nodejs.org/api/events.html#events_inheriting_from_eventemitter) shows that you'd need to call `EventEmitter.call(this);` as part of your `function Client(...)` function but I'm not sure why.

File: src/client.js [hunk 3]
```diff
@@ -73,8 +78,11 @@ function Client(sockets, name, config) {
...
 
+/** An emitter that emits both locally and through network */
 Client.prototype.emit = function(event, data) {
+	EventEmitter.prototype.emit.apply(this, arguments);
```
Comment (astorije) [inline src/client.js:86]: If `Client` inherits from `EventEmitter`, shouldn't [`this.emit(this, arguments);`](https://nodejs.org/api/events.html#events_emitter_emit_event_arg1_arg2) be enough?

Where did you find this line? I feel like such way of doing so would be necessary if `EventEmitter` would be something completely external to `Client`, that is, if there was no inheritance mechanism between those two.

If I'm right, then only one is needed, the ugly `prototype.emit.apply` call vs. the overkill-looking inheritance mechanism. If I'm wrong, slap me in the face.

Comment (astorije) [inline src/client.js:86]: Wait, I _am_ wrong! _Paf!_

Thanks, name collision :-) Ew

:+1:

File: src/models/OtrStore.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		privKey: undefined,
+		sessions: {},
+		client: client,
+	});
+	console.log('Generating DSA....');
+	this.privKey = new DSA();
+	console.log('OK, key generated');
+}
+
+/** Initialize an OTR session for query chans
```
Comment (astorije) [inline src/models/OtrStore.js]: Also, not sure if you were trying to be [JSDoc](http://usejsdoc.org/)-compliant, but if so the format is more:

``` js
/**
 * Description
 */
```

File: src/models/OtrStore.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+module.exports = OtrStore;
+
+/** Stores OTR state for a given user
+ *
+ * Handels private key, current sessions and fingerprints.
+ */
+function OtrStore(name, client) {
+	_.merge(this, {
+		name: name,
+		privKey: undefined,
```
Comment (astorije) [inline src/models/OtrStore.js]: ~~By default, non-existing properties of an object are already `undefined`:~~

``` js
$ node
> var s = new String;
undefined
> s
[String: '']
> s.foobar
undefined
```

~~The fact that they do not throw `ReferenceError` exceptions as do variables can be [a big pain in some cases](https://github.com/moll/js-must#asserting-on-property-access) but in your case, it means~~ you don't need this line. (**EDIT:** Read below why this explanation doesn't hold here).

File: src/models/OtrStore.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
+var _ = require("lodash");
+var DSA = require('otr').DSA;
+var OTR = require('otr').OTR;
+var Chan = require("./chan");
+
+module.exports = OtrStore;
+
+/** Stores OTR state for a given user
+ *
+ * Handels private key, current sessions and fingerprints.
```
Comment (astorije) [inline src/models/OtrStore.js]: Handels und Gretels? :) Handles maybe?

File: src/models/OtrStore.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ */
+function OtrStore(name, client) {
+	_.merge(this, {
+		name: name,
+		privKey: undefined,
+		sessions: {},
+		client: client,
+	});
```
Comment (astorije) [inline src/models/OtrStore.js]: Actually, this is gross. Please do instead:

``` js
this.name = name;
this.sessions = {};
this.client = client;
```

I would imagine the benefit of using [`_.merge`](https://lodash.com/docs#merge) is to be concise, but we even save 2 lines here...
Apart from that, it's useful when you don't know the keys, or when objects are big, or when deep-nesting is an issue... None of which can ever happen here.

File: src/models/OtrStore.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	console.log('Generating DSA....');
+	this.privKey = new DSA();
+	console.log('OK, key generated');
```
Comment (astorije) [inline src/models/OtrStore.js]: Reminder for later: if at some point you plan to test these, you'll have to remove the `console.log`s, acting as side-effects here.

File: src/models/OtrStore.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+/** Initialize an OTR session for query chans
+ *
+ * Event handler, to be called on chan creation
+ */
+OtrStore.prototype.initChan = function(args) {
+	var network_id = args.network;
+	var chan = args.chan;
+	if (chan.type == Chan.Type.QUERY) {
```
Comment (astorije) [inline src/models/OtrStore.js]: Always use `===` instead of `==` and `!==` instead of `!=`, [as they do not do type coercion](https://stackoverflow.com/questions/359494/does-it-matter-which-equals-operator-vs-i-use-in-javascript-comparisons). It's free, has no drawbacks, and it will save your life someday. OK, maybe not your life, but still.
Some linter configurations even fail when using these...

File: src/models/OtrStore.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+		var network = _.find(this.client.networks, {id: network_id});
+		var session = this.getSession(chan.name, network);
+		var irc = network.irc;
+		if (session == undefined) {
+			session = new OTR({
+				fragment_size: 140,
+				send_interval: 200,
```
Comment (astorije) [inline src/models/OtrStore.js]: What justifies/How did you choose these values?

File: src/models/OtrStore.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+			session.on('io', function(msg, meta) {
+				irc.send(chan.name, msg);
+			});
+			this.registerSession(chan.name, network, session);
+		}
+	}
+};
+
+OtrStore.prototype.getSession = function(nick, network) {
+	var qualifiedUser = nick + '@' + network.host;
```
Comment (astorije) [inline src/models/OtrStore.js:62]: Maybe a dumb question, but what happens if the user changes nickname in the middle of the session?

File: src/models/OtrStore.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+			session.on('ui', function(msg, encrypted, meta) {
+				irc.emit("message", {
+					from: chan.name,
+					to: irc.me,
+					message: msg
+				});
+			});
+
+			session.on('io', function(msg, meta) {
```
Comment (astorije) [inline src/models/OtrStore.js:53]: I would suggest you add a short comment above each of these 2 handlers to describe when they arrive. Probably redundant of their doc but makes it clearer when reading code.

File: src/plugins/inputs/msg.js [hunk 13]
```diff
@@ -19,7 +20,12 @@ module.exports = function(network, chan, cmd, args) {
 		target = chan.name;
 	}
 	var msg = args.join(" ");
-	irc.send(target, msg);
+	var otrSession = client.otrStore.getSession(chan.name, network);
+	if (otrSession && (otrSession.msgstate == OTR.CONST.MSGSTATE_ENCRYPTED)) {
```
Comment (astorije) [inline src/plugins/inputs/msg.js]: Does `otrSession.msgstate` contains the state of the last message? A bit too stateful IMO :P
If there is a stateless way of doing so, could you do something like `this.otrStore.isEncrypted(msg)` that would encapsulate these 2 checks and prevent you to require `OTR` line 2 and retrieve the `otrSession` line 23?

File: src/plugins/irc-events/message.js [hunk 14]
```diff
@@ -10,6 +10,11 @@ module.exports = function(irc, network) {
 			return;
 		}
 
+		if (data.message.substring(0,4) == '?OTR') {
+			// Hide OTR messages
+			return;
+		}
```
Comment (astorije) [inline src/plugins/irc-events/message.js:16]: Hey look, `if (data.message.substring(0,4) == '?OTR') return;` is less than 80 characters! :D

I think I would even prefer something along the lines of `if (client.otrStore.isOtr(data.message))`, what do you think?

File: src/plugins/irc-events/otrmessage.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
+var _ = require("lodash");
+var Chan = require("../../models/chan");
+var Msg = require("../../models/msg");
+
+module.exports = function(irc, network) {
+	var client = this;
+	irc.on("message", function(data) {
+		if (data.message.substring(0,4) != '?OTR') {
+			return;
+		}
```
Comment (astorije) [inline src/plugins/irc-events/otrmessage.js:10]: Same comments as [my comment above](https://github.com/astorije/shout/pull/1/files#r39348577).

File: src/models/OtrStore.js (hunk 5 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: Well, I come from [some religion](https://www.python.org/dev/peps/pep-0020/) which doctrine says  _« explicit it better than implicit »_. I don't need this line, but it makes the whole more readable IMHO.

File: src/models/OtrStore.js (hunk 7 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: No, that's just to respect shout coding style. all initializations are done like that, so I stick to it.

File: src/models/OtrStore.js (hunk 8 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: Yep, anyway, it's not something to leave here :) But for now, I want to keep an eye on how long it takes and if the time taken is sometimes significantly longer. (as far as now, the answer is no).

File: src/models/OtrStore.js (hunk 5 above)
Comment (astorije) [inline src/models/OtrStore.js]: Mmh, forget what I said about JS oddities, it doesn't apply here anyway, my bad. Forget about being explicit or implicit as well, this has nothing to do. What you are doing here is essentially:

``` js
this.privKey = undefined;
// ...
this.privKey = new DSA();
```

This is just confusing and adds no readability: say the function goes more complex, one will wonder why the private key defaults to `undefined`.
Conclusion: either remove the line from the hash, or replace `undefined` with `new DSA();` within the hash. You can still move your `console.log()` test above that in the second solution. But one needs to go.

File: src/models/OtrStore.js (hunk 7 above)
Comment (astorije) [inline src/models/OtrStore.js]: This goes far beyond coding style, and I think I understand why someone would ever want to do that.

``` js
function F1(a) {
  this.a = a;
}

function F2(a) {
  require('lodash').merge(this, { a: a });
}

var a = { foo: 'bar' };
var a1 = new F1(a);
var a2 = new F2(a);
a.foo = 'bar2';
console.log(a1.a.foo); // bar2
console.log(a2.a.foo); // bar
```

So doing so (using `_.merge`) deeply clones all objects within the `source` hash :anguished:. Is that really what you want?!
This is really heavy, probably unnecessary in most cases, and most likely harmful in some. Don't get me wrong, I am all for immutable and persistent data structures, but here you get neither of those while potentially messing up your program if you forget objects are not shared. Nice.

By the way, I don't recall seeing any project using lodash/underscore instantiating objects with `merge`, _ever_.
Bottom line is: even if it was only for style (which it is not!), one should not spend half an hour understanding a damn object assignment.

File: src/models/OtrStore.js (hunk 8 above)
Comment (astorije) [inline src/models/OtrStore.js]: Re: temporary benchmarking, gotcha! And I figured that was your intent.
Re: time taken, great news!

File: src/client.js (hunk 2 above)
Comment (JocelynDelalande) [inline src/client.js:82]: Well, for inheritance consistency :)

File: src/models/OtrStore.js (hunk 4 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: [Not necessarily](http://usejsdoc.org/about-getting-started.html#adding-documentation-comments-to-your-code) (EDIT: but your remark is relevant here, fixing)

File: src/models/OtrStore.js (hunk 5 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: Later, I would like to generate the DSA key oportunisticly (it adds delay on shout startup currently), so it will start as undefined, and being populated later (or never, if no OTR is used).

File: src/models/OtrStore.js (hunk 6 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: :)

File: src/models/OtrStore.js (hunk 7 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: I leave it as-is, once again, for consistency, I agree your point, but that change takes place in a global pull request removing all those `_.merge()` in constructors.

File: src/models/OtrStore.js (hunk 9 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: thanks :) (delightful as PHP ;-))

File: src/models/OtrStore.js (hunk 10 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: For `fragment_size`, that's the only size I found working (even with [node-otr4 lib](https://github.com/mnaamani/node-otr4)), for `send_interval`, it was from the otr lib README.md but testing a bit more showed me that a lower value can fit and makes the whole thing faster. Thanks for questioning about that :)

File: src/models/OtrStore.js (hunk 11 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js:62]: not that dumb, but we will have to figure out what to do regarding the protocol, for example, irssi implementation regenerate a key if you change your nick (key is attached to nick).

File: src/plugins/inputs/msg.js (hunk 13 above)
Comment (JocelynDelalande) [inline src/plugins/inputs/msg.js]: > Does otrSession.msgstate contains the state of the last message?

no, more like the current "mode" : chatting encrypted or chatting cleartext.

File: src/plugins/irc-events/message.js (hunk 14 above)
Comment (JocelynDelalande) [inline src/plugins/irc-events/message.js:16]: yep

Comment (JocelynDelalande) [inline src/plugins/irc-events/message.js:16]: btw, hating single-line without curly braces, makes code less readable and not diff-friendly when evolving

File: src/client.js (hunk 2 above)
Comment (astorije) [inline src/client.js:82]: -__-'

My point was that I don't know what that call exactly does, but sure if, to the question "Why do we need this function to be inheritance-wise consistent?", the answer "for inheritance consistency" is good enough for you, be my guest :D

The comment behind my link says `// Initialize necessary properties from`EventEmitter`in this instance` so I guess we should set it anyway, but it's nice to know what an impure function does to your object when you call it and what do we risk if we forget it.

Therefore I let you pick if we need it or not and you'll have my :+1: here.

File: src/models/OtrStore.js (hunk 5 above)
Comment (astorije) [inline src/models/OtrStore.js]: > Later, I would like to [...]

That will be later. Please remove one of them for that PR :-)

> [...] generate the DSA key oportunisticly (it adds delay on shout startup currently)

Ouch.
Opportunistically or not, this should be an asynchronous call IMO (for a future PR, synchronous here is fine). A promise, perhaps... :-)

**Comment by JocelynDelalande** [comment]
Reworked the PR with your latest comments. It's not exempt of bugs (see #3 and #5) but I think that's enough to merge it in our dev branch and thus have a common base to work on.

File: src/models/OtrStore.js (hunk 7 above)
Comment (astorije) [inline src/models/OtrStore.js]: Then, knowing that you deeply clone the `client` here, while you also deeply clone the `otrStore` [there](https://github.com/astorije/shout/pull/1/files#diff-cf27c1d543e886c89cd9ac8b8aeaf05bR66) and that, therefore, you allow yourself to a world of pain (`client.otrStore.client` and `client` references 2 different instances, same for `otrStore.client.otrStore` and `otrStore`...), you will have to live with my :-1:, do what you want with it :-).

File: src/models/OtrStore.js (hunk 8 above)
Comment (astorije) [inline src/models/OtrStore.js]: :+1:

File: src/models/OtrStore.js (hunk 10 above)
Comment (astorije) [inline src/models/OtrStore.js]: How I like such methods! What units are these? What do they do? Could you maybe add a comment about these?

File: src/models/OtrStore.js (hunk 11 above)
Comment (astorije) [inline src/models/OtrStore.js:62]: Fair enough, then I guess this is :+1: for the scope of this PR, with a note that we'll have more to do in the future.

File: src/plugins/inputs/msg.js (hunk 13 above)
Comment (astorije) [inline src/plugins/inputs/msg.js]: I see, then I think I'd be happy with a helper on the `OtrStore` side of things to wrap all of these. But here as well it can be done later. :+1: for now with note to move intelligence over there, unless you were planning to do so (i.e. to get rid of your `FIXME`) beforehand. Up to you.

File: src/plugins/irc-events/message.js (hunk 14 above)
Comment (astorije) [inline src/plugins/irc-events/message.js:16]: "yep" was enough :D
I'm fine with anything (wouldn't want to be the one accused of [bikeshedding](https://en.wikipedia.org/wiki/Parkinson%27s_law_of_triviality), would I?), even with using the hardcoded `?OTR` here rather than a helper for this PR, so :+1: with note to move OTR intelligence over there.

File: package.json (hunk 1 above)
Comment (astorije) [inline package.json]: :+1:

File: src/models/OtrStore.js [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+	var chan = args.chan;
+	if (chan.type === Chan.Type.QUERY) {
+		var network = _.find(this.client.networks, {id: network_id});
+		var session = this.getSession(chan.name, network);
+		var irc = network.irc;
+		if (session === undefined) {
+			session = new OTR({
+				/* After some tests, the only working value  with this lib and
+				 * node-otr4 lib as well
+				 */
```
Comment (astorije) [inline src/models/OtrStore.js]: Call me relou but when inline commenting, no need for that extra line, please go with:

``` js
// After some tests, the only working value  with this lib and
// node-otr4 lib as well
```

File: src/models/OtrStore.js [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+				fragment_size: 140,
+
+				/* after some tests, a good value not get kicked, and to be
+				 * fast enough, at least on FreeNode.
+				 * (tested pasting shout README, 742 chars).
+				 */
```
Comment (astorije) [inline src/models/OtrStore.js]: I would actually comment on what the hell these values are/what these values are doing, not that "after some tests, these seem good" :D

File: src/models/OtrStore.js [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+		}
+	}
+};
+
+OtrStore.prototype.getSession = function(nick, network) {
+	var qualifiedUser = nick + '@' + network.host;
+	return this.sessions[qualifiedUser];
+};
+
+
```
Comment (astorije) [inline src/models/OtrStore.js]: ... One extra new line... :kissing_closed_eyes: 

[_Finish him!_](https://www.youtube.com/watch?v=_hHDxlm66dE)

File: src/models/OtrStore.js [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * Stores OTR state for a given user
+ *
+ * Handles current sessions.
+ */
+function OtrStore(name, client) {
+	_.merge(this, {
+		name: name,
+		sessions: {},
+		client: client,
```
Comment (astorije) [inline src/models/OtrStore.js]: Trailing comma here.

File: src/models/OtrStore.js (hunk 7 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: I don't see where does the deep clone takes place in all that :-/ where is code/doc about that ?

Comment (astorije) [inline src/models/OtrStore.js]: Doc is [here](https://lodash.com/docs#merge) and proof is [here](https://github.com/astorije/shout/pull/1#discussion_r39352156).

File: src/models/OtrStore.js (hunk 10 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: > Could you maybe add a comment about these?

I did, actually (explaining why those values). BTW I'm not going to do all otr lib documentation here.

Being kicked of an IRC network for excessive flood is not science, so I see no better choice than experimentation.

(btw, no unit cause they were obvious imho : a time in network protocol is milliseconds, and a message size on a char-based protocol is in chars).

**Comment by astorije** [comment]
I'll look into #3 before the final :+1:/merging as it might be more problematic than it looks.

**Comment by JocelynDelalande** [comment]
> I'll look into #3 before the final :+1:/merging as it might be more problematic than it looks.

Please don't, that's PR to a dev branch, it can have issues (having this hanging PR is quite blocking), basic functionality is achieved to allow continuing development (and fixing them).

File: src/models/OtrStore.js (hunk 5 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js]: For what I was doing anyway I didn't even needed to generate the key by hand, short-lived keys are built on-demand by the OTR lib if none is given in `OTR()` options. So I removed everything about that privKey for that first PR.

File: src/models/OtrStore.js (hunk 10 above)
Comment (astorije) [inline src/models/OtrStore.js]: Sorry, maybe I didn't explain myself properly. Reading the code as someone not involved in this file, it doesn't strike me as obvious. Without writing all of the lib doc here, understanding the code at first glance might vital here, especially if someone reviews the final code to merge it with the upstream and doesn't want to dig into the lib.

But it's okay, we have more pressing concerns at the moment. I'd like to do an entire pass of documenting the overall code myself anyway, before sending the full PR.

:+1:

File: src/models/OtrStore.js (hunk 4 above)
Comment (astorije) [inline src/models/OtrStore.js]: :+1:

File: src/models/OtrStore.js (hunk 6 above)
Comment (astorije) [inline src/models/OtrStore.js]: :+1:

File: src/models/OtrStore.js (hunk 12 above)
Comment (JocelynDelalande) [inline src/models/OtrStore.js:53]: done

File: src/models/OtrStore.js (hunk 5 above)
Comment (astorije) [inline src/models/OtrStore.js]: :+1:

File: src/plugins/inputs/msg.js (hunk 13 above)
Comment (JocelynDelalande) [inline src/plugins/inputs/msg.js]: Will fix later.

File: src/client.js [hunk 20]
```diff
@@ -51,15 +55,17 @@ var inputs = [
...
+	var client = this;
 	_.merge(this, {
 		activeChannel: -1,
 		config: config,
 		id: id++,
 		name: name,
 		networks: [],
-		sockets: sockets
+		sockets: sockets,
+		otrStore: new OtrStore(name, client),
```
Comment (astorije) [inline src/client.js]: Trailing comma here.

File: src/plugins/irc-events/otrmessage.js [hunk 21]
```diff
@@ -0,0 +1,14 @@
+var _ = require("lodash");
+var Chan = require("../../models/chan");
+var Msg = require("../../models/msg");
+
+module.exports = function(irc, network) {
+	var client = this;
+	irc.on("message", function(data) {
+		if (! client.otrStore.isOtrMessage(data.message)) {
```
Comment (astorije) [inline src/plugins/irc-events/otrmessage.js]: Well well well, look who's there, an unexpected whitespace! :-)

**Comment by astorije** [comment]
@JocelynDelalande, my dear friend, you hereby get my :+1: :-)

**Comment by JocelynDelalande** [comment]
> @JocelynDelalande, my dear friend, you hereby get my :+1: :-)

gosh! it's a pleasure to nitpick with you ;)
